doc/man/man1/qrstat.1
'\" t
.\"___INFO__MARK_BEGIN__
.\"
.\" Copyright: 2004-2007 by Sun Microsystems, Inc.
.\"
.\"___INFO__MARK_END__
.\"
.\"
.\" Some handy macro definitions [from Tom Christensen's man(1) manual page].
.\"
.de SB		\" small and bold
.if !"\\$1"" \\s-2\\fB\&\\$1\\s0\\fR\\$2 \\$3 \\$4 \\$5
..
.\" "
.de T		\" switch to typewriter font
.ft CW		\" probably want CW if you don't have TA font
..
.\"
.de TY		\" put $1 in typewriter font
.if t .T
.if n ``\c
\\$1\c
.if t .ft P
.if n \&''\c
\\$2
..
.\"
.de M		\" man page reference
\\fI\\$1\\fR\\|(\\$2)\\$3
..
.TH QRSTAT 1 "2011-05-20" "xxRELxx" "xxQS_NAMExx User Commands"
.SH NAME
qrstat \- show the status of xxQS_NAMExx Advance Reservations (AR)
.\"
.\"
.SH SYNTAX
.B qrstat
.RB [ \-ar
.IR ar_id ,...]
.RB [ \-help ]
.RB [ \-u
.IR user ,...]
.RB [ \-explain ]
.\"
.\"
.SH DESCRIPTION
.I qrstat
shows the current status of the available xxQS_NAMExx ARs.
The selection option \fB\-ar\fP allows you to get information about 
specific AR.
.PP
The administrator and the user may define files
which can contain any of the options described below. A cluster-wide sge_qrstat
file may be placed under
$xxQS_NAME_Sxx_ROOT/$xxQS_NAME_Sxx_CELL/common/sge_qrstat
The user private  file is searched at the location
$HOME/.sge_qrstat.
The home directory request file has the highest precedence over
the cluster global file.
Command line can be used to override the flags contained in the files.
.\"
.SH OPTIONS
.IP "\fB\-ar\fP \fIar_id\fP,..."
Prints various information about the ARs identified by the given \fIar_id\fP list.
.\"
.IP "\fB\-explain\fP"
Displays the reason for the error state of an AR. Possible reasons 
are the unknown state of a host or queue instance.
.sp
The output format for the alarm reasons is one line per reason.
.\"
.IP "\fB\-help\fP"
Prints a listing of all options.
.\"
.IP "\fB\-u\fP \fIuser\fP,..."
Display information only for those ARs 
created by the users from the given user list.
.sp
The string
\fB\ $user\fP
is a placeholder for the current username. An asterisk "\fB*\fB" can be used
as a username wildcard to request that all users' ARs be displayed. The
default value for this switch is "\fB-u $user\fP".
.\"
.IP "\fB\-xml\fP"
This option can be used with all other options and changes the output to XML. The schemas
used are referenced in the XML output. The output is printed to stdout.
.br
.\"
.\"
.SH "OUTPUT FORMATS"
Depending on the presence or absence of the \fB\-ar\fP option there are two 
output formats need to be differentiated.
.\"
.SS "\fBAdvance Reservation Summary (without \-ar)\fP"
Following the header line, a section for each AR
is provided. The columns contain information for
.IP "\(bu" 3n
the AR id;
.IP "\(bu" 3n
the name of the AR;
.IP "\(bu" 3n
the current state of the AR. One of the following letters with the
given meaning:
.sp 1
.nf
.ta \w'|b|  'u
\fBw\fP \- waiting without error
\fBW\fP \- warning (effective \- waiting with error)
\fBr\fP \- running
\fBE\fP \- error (effective \- running with error)
\fBd\fP \- deleted
.fi
.IP "\(bu" 3n
the start time of the AR;
.IP "\(bu" 3n
the end time of the AR;
.IP "\(bu" 3n
the duration of the AR.
.SS "\fBDetailed Format (with \-ar)\fP"
The output contains two columns. The first one contains all AR attributes.
The second one the corresponding value.
.\"
.\"
.SH "ENVIRONMENT VARIABLES"
.\" 
.IP "\fBxxQS_NAME_Sxx_ROOT\fP" 1.5i
Specifies the location of the xxQS_NAMExx standard configuration
files.
.\"
.IP "\fBxxQS_NAME_Sxx_CELL\fP" 1.5i
If set, specifies the default xxQS_NAMExx cell. To address a xxQS_NAMExx
cell
.I qrstat
uses (in the order of precedence):
.sp 1
.RS
.RS
The name of the cell specified in the environment 
variable xxQS_NAME_Sxx_CELL, if it is set.
.sp 1
The name of the default cell, i.e. \fBdefault\fP.
.sp 1
.RE
.RE
.\"
.IP "\fBxxQS_NAME_Sxx_DEBUG_LEVEL\fP" 1.5i
If set, specifies that debug information
should be written to stderr. In addition the level of
detail in which debug information is generated is defined.
.\"
.IP "\fBxxQS_NAME_Sxx_QMASTER_PORT\fP" 1.5i
If set, specifies the tcp port on which
.M xxqs_name_sxx_qmaster 8
is expected to listen for communication requests.
Most installations will use a services map entry for the
service "sge_qmaster" instead to define that port.
.\"
.\"
.SH FILES
.nf
.ta \w'<xxqs_name_sxx_root>/     'u
\fI<xxqs_name_sxx_root>/<cell>/common/act_qmaster\fP
	xxQS_NAMExx master host file
.ta \w'<xxqs_name_sxx_root>/     'u
\fI<xxqs_name_sxx_root>/<cell>/common/xxqs_name_sxx_qrstat\fP
	cluster qrstat default options
\fI$HOME/.xxqs_name_sxx_qrstat\fP	
	user qrstat default options
.fi
.\"
.\"
.SH "SEE ALSO"
.M xxqs_name_sxx_intro 1 ,
.M qrsub 1 ,
.M qrdel 1 ,
.M qsub 1 ,
.\"
.\"
.SH "COPYRIGHT"
See
.M xxqs_name_sxx_intro 1
for a full statement of rights and permissions.


NEWS
Notable changes in Son of Grid Engine
=====================================

For earlier changes (comprising Sun Grid Engine), see Changelog in the
source distribution, and the bug lists at
http://arc.liv.ac.uk/SGE/howto/Installation,%20Upgrade,%20Patches.
For detailed recent change information, including credits for changes
not listed here, see https://arc.liv.ac.uk/trac/SGE/log/sge/

"[#<n>]" below refers to ticket <n> at
https://arc.liv.ac.uk/trac/SGE/ticket/<n>.  For historical reasons,
changes due to Univa/Sun are tagged as "[U]", or "[(U)]" if partially
due to them or appearing first in this version.


Version 8.1.9
-------------

* Security
  * Drop setgid privileges correctly (CERT POS36-C); daemons now run
    with (only) admin user's groups, not root's

* Building

  * MUNGE authentication can be built in (aimk -munge)
  * Use a minimal autoconf framework
  * Fix building on recent Darwin and Cygwin [#1516, #1536, #1557]
  * Drop support for Darwin < 10 and Darwin/PPC
  * Cope with GNU/Linux ppc64le as well as ppc64el
  * Fix rpm packaging of spool objects [#1519]
  * Actually install "work" in utilbin (c.f. 8.1.8)
  * Rpm spec fixes for Fedora 23, ppc64, and OpenSuSE
  * Fix build when bool is defined in headers

* Bug fixes

  * Man fixes
  * Message fixes [#1538, #1539]
  * Fix RPM packaging of spool files [#1519]
  * Don't set port and cluster in check_sge.py, and accept memory in T
  * Fix pam_sge-qrsh-setup for recent PAM
  * Do a full, not a partial match on GECOS data when selecting certificate
    in sge_ca [#1517]
  * Fix checking of admin host in qconf [#1547]
  * Report the actual exit status of master task in tightly integrated
    parallel jobs
  * Fix share tree accounting with array jobs [#435]
  * Fix doubled multipliers of units in qacct memory reports
  * Improve job scheduling within a sharetree node [#1550]
  * Flush spool at qmaster exit [#1551]
  * Sort jobs and advance reservations when reading from spool [#1552, #1553]
  * Fix detection of old v. new type of cpuset [#1523]
  * Fix treatment of some invalid entries in variable lists or the
    environment on job submission

* Enhancements

  * MUNGE authentication (as above)
  * qconf -clearusage can act on a specific user
  * Allow generating load sensor data just before sending them
  * Add sge_ca "-revoke <cert>" option
  * Allow running DRMAA jobs as arbitrary user [#492].  (Partially
    done in v 8.1.0, but not mentioned; now works with CSP.)
  * Report SSL error numbers in hex [#1539]
  * Use PVM_VMID environment variable [#44]
  * rc templates support a "status" argument (for HA usage)

* Potentially incompatible changes

  * Print error messages to stderr and don't print on success with -terse
    to be consistent with submit(1) saying that error messages go to
    stderr and to make -terse more useful in cases like
      cat test.o$(qsub -sync y -terse -cwd -N test ...)
  * sge_ca -renew and friends no longer revoke old certificates
  * The qconf -sep and -cb options have been removed finally
  * A POSIX 2001-compatible library is required

Version 8.1.8
-------------

* Building

  * Fix build with recent GNU binutils
  * Fix rpm build on Fedora and maybe use packaged swing-layout
  * Fix dpkg build on recent Debian/Ubuntu
  * uid_t/gid_t are unsigned on NetBSD
  * Install "work" in utilbin
  * Move qsched back to main package

* Bug fixes

  * Man fixes
  * Fix buffer overflow in qmake [#1507]
  * Avoid possible link-time error (Debian bug # 749413)
  * Fix some global messages from qstat -j
  * Avoid truncating usage values in accounting records
  * Fix execd remembering existing core binding when restarted [#1511]

* Enhancements

  * Nagios and Ganglia monitoring scripts
  * Python JSV implementation
  * Upstart and systemd startup files
  * Define DRMAA_LIBRARY_PATH in setup scripts

Version 8.1.7
-------------

* Building

  * Remove the shipped old jemalloc in favour of an external one
    (aimk -jemalloc)
  * Always use system libraries -- never load SSL stuff dynamically
  * aimk no longer uses specific paths for openssl etc.  Use
    INPUT_*_FLAGS if necessary
  * SUpport building rpm on OpenSuSE and RHEL7

* Bug fixes

  * Man, other doc, and message fixes/changes [#1504]
  * Fix minor memory leaks and check result of all malloc calls
  * Don't pause for a second with client JSVs [#1475]
  * Allow longer "failed" strings in admin mail
  * Fix spurious thread error reports from qping [#588]
  * Avoid possible crash reading malformed configuration files
  * Fix GUI installer startup [#1488]
  * Always define SGE_STARTER_USE_LOGIN_SHELL with starter method
  * Modify cpuset-based process killing [#1483, #1490]
  * Make various files optional when setting permissions at installation
  * Fix cpu-bound shepherd with qlogin
  * Fix queue list for adding queue in qmon [#1495]
  * Clean up some file permissions and shebangs in non-executables [#1493]
  * Really ignore -e, -o args of qrsh/qsh/qlogin, as documented
  * Fix use of SGE_QMASTER_PIDFILE and double writing of pid file
  * Fix occasional qmaster major space leak with schedd_job_info [#360, #682]
  * Wait until daemon goes down when stopping sge_execd

* Enhancements

  * Support Cygwin x86_64, and GNU/Linux on arm64 and ppc64el (untested)
  * qconf -aattr et al now support modifying types user, userset, calendar,
    and project
  * Translate exit code to signal name in qacct output
  * qacct -m prints only master tasks
  * Add ar_sub_time to qacct output
  * Allow rctemplate configuration from /etc/sysconfig as well as
    /etc/default
  * Update MPI PE templates and install a tight integration "mpi" PE
  * Make suid files, dtrace, qmon, start_gui_installer optional in
    setfileperm.sh (used at installation)
  * Speed up JSV in bash by not calling cut

* Potentially incompatible changes

  * Default ENABLE_ADDGRP_KILL to on
  * Clean symbol exports from libdrmaa.so (GNU ld only); now only exports

source/clients/qrstat/qrstat.c
/*___INFO__MARK_BEGIN__*/
/*************************************************************************
 * 
 *  The Contents of this file are made available subject to the terms of
 *  the Sun Industry Standards Source License Version 1.2
 * 
 *  Sun Microsystems Inc., March, 2001
 * 
 * 
 *  Sun Industry Standards Source License Version 1.2
 *  =================================================
 *  The contents of this file are subject to the Sun Industry Standards
 *  Source License Version 1.2 (the "License"); You may not use this file
 *  except in compliance with the License. You may obtain a copy of the
 *  License at http://gridengine.sunsource.net/Gridengine_SISSL_license.html
 * 
 *  Software provided under this License is provided on an "AS IS" basis,
 *  WITHOUT WARRANTY OF ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING,
 *  WITHOUT LIMITATION, WARRANTIES THAT THE SOFTWARE IS FREE OF DEFECTS,
 *  MERCHANTABLE, FIT FOR A PARTICULAR PURPOSE, OR NON-INFRINGING.
 *  See the License for the specific provisions governing your rights and
 *  obligations concerning the Software.
 * 
 *   The Initial Developer of the Original Code is: Sun Microsystems, Inc.
 * 
 *   Copyright: 2001 by Sun Microsystems, Inc.
 * 
 *   All Rights Reserved.
 * 
 ************************************************************************/
/*___INFO__MARK_END__*/

#include "uti/sge_rmon.h"
#include "uti/sge_bootstrap.h"
#include "uti/sge_uidgid.h"
#include "uti/sge_io.h"
#include "uti/sge_string.h"
#include "uti/sge_log.h"

#include "gdi/sge_gdi.h"
#include "gdi/sge_gdi_ctx.h"
#include "gdi/sge_gdi2.h"

#include "sgeobj/sge_all_listsL.h"
#include "sgeobj/parse.h"
#include "sgeobj/sge_answer.h"
#include "sgeobj/sge_str.h"

#include "basis_types.h"
#include "sge.h"
#include "read_defaults.h"
#include "parse_qsub.h"
#include "usage.h"
#include "sig_handlers.h"
#include "qrstat.h"
#include "qrstat_report_handler.h"
#include "qrstat_report_handler_xml.h"
#include "qrstat_report_handler_stdout.h"
#include "qrstat_filter.h"
#include "msg_common.h"

extern char **environ;

static bool
sge_parse_from_file_qrstat(const char *file, lList **ppcmdline, lList **alpp);

static bool 
sge_parse_qrstat(sge_gdi_ctx_class_t *ctx, lList **answer_list,
                 qrstat_env_t *qrstat_env, lList **cmdline)
{
   bool ret = true;
   
   DENTER(TOP_LAYER, "sge_parse_qrstat");

   qrstat_env->is_summary = true;
   while (lGetNumberOfElem(*cmdline)) {
      u_long32 value;
   
      /* -help */
      if (opt_list_has_X(*cmdline, "-help")) {
         sge_usage(QRSTAT, stdout);
         DEXIT;
         SGE_EXIT((void**)&ctx, 0);
      }

      /* -u */
      while (parse_multi_stringlist(cmdline, "-u", answer_list, 
                                    &(qrstat_env->user_list), ST_Type, ST_name)) {
         continue;
      }

      /* -explain */
      while (parse_flag(cmdline, "-explain", answer_list, &value)) {
         qrstat_filter_add_core_attributes(qrstat_env);
         qrstat_filter_add_explain_attributes(qrstat_env);
         qrstat_env->is_explain = (value > 0) ? true : false;
         continue;
      }

      /* -xml */
      while (parse_flag(cmdline, "-xml", answer_list, &value)) {
         qrstat_filter_add_core_attributes(qrstat_env);
         qrstat_filter_add_xml_attributes(qrstat_env);
         qrstat_env->is_xml = (value > 0) ? true : false;
         continue;
      }

      /* -ar */
      while (parse_u_longlist(cmdline, "-ar", answer_list, &(qrstat_env->ar_id_list))) {         
         qrstat_filter_add_core_attributes(qrstat_env);
         qrstat_filter_add_ar_attributes(qrstat_env);
         qrstat_filter_add_ar_where(qrstat_env);
         qrstat_env->is_summary = false;
         continue;      
      }

      if (lGetNumberOfElem(*cmdline)) {
         sge_usage(QRSTAT, stdout);
         answer_list_add(answer_list, MSG_PARSE_TOOMANYOPTIONS, 
                         STATUS_ESEMANTIC, ANSWER_QUALITY_ERROR);
         ret = false;
         break;
      }
   } 

   if (qrstat_env->is_summary) {
      char user[128] = "";
      if (sge_uid2user(geteuid(), user, sizeof(user), MAX_NIS_RETRIES)) {
         answer_list_add_sprintf(answer_list, STATUS_ESEMANTIC,
                                 ANSWER_QUALITY_CRITICAL,
                                 MSG_SYSTEM_RESOLVEUSER_U, (u_long32)geteuid());
         ret = false;
      } else {
         str_list_transform_user_list(&(qrstat_env->user_list), answer_list, user);
         qrstat_filter_add_core_attributes(qrstat_env);
         qrstat_filter_add_u_where(qrstat_env);
      }
   }

   DRETURN(ret);
}

/************************************************************************/
int main(int argc, char **argv) {
   int ret = 0;
   lList *pcmdline = NULL;
   lList *answer_list = NULL;
   sge_gdi_ctx_class_t *ctx = NULL;
   qrstat_env_t qrstat_env;

   DENTER_MAIN(TOP_LAYER, "qrsub");

   /* Set up the program information name */
   sge_setup_sig_handlers(QRSTAT);

   log_state_set_log_gui(1);

   if (sge_gdi2_setup(&ctx, QRSTAT, MAIN_THREAD, &answer_list) != AE_OK) {
      answer_list_output(&answer_list);
      goto error_exit;
   }

   qrstat_filter_init(&qrstat_env);
   qrstat_filter_set_ctx(&qrstat_env, ctx);

   /*
    * stage 1: commandline parsing
    */
   {
      dstring file = DSTRING_INIT;
      const char *user = ctx->get_username(ctx);
      const char *cell_root = ctx->get_cell_root(ctx);

      /* arguments from SGE_ROOT/common/sge_qrstat file */
      get_root_file_path(&file, cell_root, SGE_COMMON_DEF_QRSTAT_FILE);
      if (sge_parse_from_file_qrstat(sge_dstring_get_string(&file), &pcmdline, &answer_list) == true) {
         /* arguments from $HOME/.sge_qrstat file */
         if (get_user_home_file_path(&file, SGE_HOME_DEF_QRSTAT_FILE, user, &answer_list)) {
            sge_parse_from_file_qrstat(sge_dstring_get_string(&file), &pcmdline, &answer_list);
         }
      }
      sge_dstring_free(&file); 

      if (answer_list) {
         answer_list_output(&answer_list);
         lFreeList(&pcmdline);
         sge_prof_cleanup();
         SGE_EXIT((void**)&ctx, 1);
      }
   }

   answer_list = cull_parse_cmdline(QRSTAT, argv+1, environ, &pcmdline, FLG_USE_PSEUDOS);
   if (answer_list != NULL) {
      answer_list_output(&answer_list);
      lFreeList(&pcmdline);
      goto error_exit;
   }
  
   /* 
    * stage 2: evalutate switches and modify qrstat_env
    */
   if (!sge_parse_qrstat(ctx, &answer_list, &qrstat_env, &pcmdline)) {
      answer_list_output(&answer_list);
      lFreeList(&pcmdline);
      goto error_exit;
   }

   /* 
    * stage 3: fetch data from master 
    */
   {
      answer_list = ctx->gdi(ctx, SGE_AR_LIST, SGE_GDI_GET, &qrstat_env.ar_list, 
                     qrstat_env.where_AR_Type, qrstat_env.what_AR_Type, false);

      if (answer_list_has_error(&answer_list)) {
         answer_list_output(&answer_list);
         goto error_exit;
      }
   }

   /*
    * stage 4: create output in correct format
    */
   {
      qrstat_report_handler_t *handler = NULL;

      if (qrstat_env.is_xml) {
         handler = qrstat_create_report_handler_xml(&qrstat_env, &answer_list);
      } else {
         handler = qrstat_create_report_handler_stdout(&qrstat_env, &answer_list);
      }
      if (!qrstat_print(&answer_list, handler, &qrstat_env)) {
         ret = 1;
      }
      if (qrstat_env.is_xml) {
         qrstat_destroy_report_handler_xml(&handler, &answer_list); 
      } else {
         qrstat_destroy_report_handler_stdout(&handler, &answer_list); 
      }
   }

   sge_gdi2_shutdown((void**)&ctx);
   sge_prof_cleanup();
   DRETURN(ret);

error_exit:
   sge_gdi2_shutdown((void**)&ctx);
   sge_prof_cleanup();
   SGE_EXIT((void**)&ctx, 1);
   DRETURN(1);
}

static bool
sge_parse_from_file_qrstat(const char *file, lList **ppcmdline, lList **alpp)
{
   bool ret = true;

   DENTER(TOP_LAYER, "sge_parse_from_file_qrstat");

   if (ppcmdline == NULL) {
      ret = false;
   } else {
      if (!sge_is_file(file)) {
         /*
          * This is no error
          */
         DPRINTF(("file "SFQ" does not exist\n", file));
      } else {
         char *file_as_string = NULL;
         int file_as_string_length;

         file_as_string = sge_file2string(file, &file_as_string_length);
         if (file_as_string == NULL) {
            answer_list_add_sprintf(alpp, STATUS_EUNKNOWN, 
                                    ANSWER_QUALITY_ERROR,
                                    MSG_ANSWER_ERRORREADINGFROMFILEX_S, file);
            ret = false;
         } else {
            char **token = NULL;

            token = stra_from_str(file_as_string, " \n\t");
            *alpp = cull_parse_cmdline(QRSTAT, token, environ, ppcmdline, FLG_USE_PSEUDOS);
         }
      }
   }  
   DRETURN(ret); 
}
